### ISSUE TITLE
TimSort is broken on Julia v0.4

### ISSUE THREAD

**Issue by kmsquire** [JuliaLang/SortingAlgorithms.jl#8]
```julia
julia> x = rand(100000);

julia> sort(x, alg=TimSort)
ERROR: BoundsError()
 in merge_hi at /Users/kevin/.julia/v0.4/SortingAlgorithms/src/SortingAlgorithms.jl:445
 in merge at /Users/kevin/.julia/v0.4/SortingAlgorithms/src/SortingAlgorithms.jl:336
 in merge_collapse at /Users/kevin/.julia/v0.4/SortingAlgorithms/src/SortingAlgorithms.jl:305
 in sort! at /Users/kevin/.julia/v0.4/SortingAlgorithms/src/SortingAlgorithms.jl:558
 in fpsort! at sort.jl:457
 in sort! at sort.jl:461
 in sort! at sort.jl:342
 in sort at sort.jl:346
```

Seems to work on v0.3.

**Comment by mlhetland** [comment]
Maybe worth pointing out that TimSort seems to be [broken in Python and Java](http://envisage-project.eu/proving-android-java-and-python-sorting-algorithm-is-broken-and-how-to-fix-it/) as well. Not sure if this bug has snuck into Julia?

**Comment by kmsquire** [comment]
Thanks for the link (and for bumping this thread)--yes, this is broken in Julia as well, since we use the same invariant.

**Comment by hayd** [comment]
It's kind of an [annoying game of spot the difference](http://envisage-project.eu/proving-android-java-and-python-sorting-algorithm-is-broken-and-how-to-fix-it/#sec3.1), but it looks like it requires a [4 runs to be checked rather than 3](https://github.com/JuliaLang/SortingAlgorithms.jl/blob/06f6d8f41d2cf350be87a9fa1d7a62dfcc46c230/src/SortingAlgorithms.jl#L293)... It's unclear to me but I think this means you can just change the subsequent line to something like (untested):

    if length(a) > length(b)+length(c) && length(b) > length(c) && (length(state.runs) > 3) && length(state.runs[end-3]) > length(a)+length(b)) && !force

*(Note: may be worth defining `length(state.runs)` in the while loop condition and using `@inbounds`, not sure.)*

**Comment by kmsquire** [comment]
So, it turns out that `TimSort` isn't broken on Julia, at least not in a way that is important:

1. In my original bug report, I had checked out master, which is incorrectly pointing at v0.0.2 of SortingAlgorithms.jl.  The latest version (0.0.4) works fine.  (I'll fix master).
2. From the link discussing why timsort is broken in Python and Java, the reason it's broken there is that timsort keeps an array of sorted subarray indices, and because of a mistaken invariant, could overflow the (fixed size) array allocated in both languages with a.  In the implementation here, we used a variable sized array, and so would never run into that problem (the max array size needed for the "broken" version in Java is 49). 

The implementation is still wrong, of course, in the sense that the original invariant tests are insufficient--I'll commit a fix shortly.
